File: db/src/main/java/dev/qixils/aviary/db/DatabaseManager.java
package dev.qixils.aviary.db;

import com.mongodb.ConnectionString;
import com.mongodb.MongoClientSettings;
import com.mongodb.client.model.Filters;
import com.mongodb.reactivestreams.client.MongoClient;
import com.mongodb.reactivestreams.client.MongoClients;
import com.mongodb.reactivestreams.client.MongoCollection;
import com.mongodb.reactivestreams.client.MongoDatabase;
import org.bson.codecs.configuration.CodecRegistries;
import org.bson.codecs.configuration.CodecRegistry;
import org.bson.codecs.pojo.PojoCodecProvider;
import org.bson.conversions.Bson;
import org.jetbrains.annotations.NotNull;
import reactor.core.publisher.Flux;

import java.io.Closeable;
import java.util.Locale;
import java.util.Map;

/**
 * Manages the connection to a MongoDB database.
 */
public class DatabaseManager implements Closeable {
	private final @NotNull CodecRegistry pojoRegistry = CodecRegistries.fromRegistries(MongoClientSettings.getDefaultCodecRegistry(),
			CodecRegistries.fromProviders(PojoCodecProvider.builder().automatic(true).build()));
	private final @NotNull MongoClientSettings mongoClientSettings = MongoClientSettings.builder()
			.applyConnectionString(new ConnectionString("mongodb://localhost:27017"))
			.codecRegistry(pojoRegistry)
			.build();
	private final @NotNull MongoClient mongoClient = MongoClients.create(mongoClientSettings);
	protected final @NotNull MongoDatabase database;

	public DatabaseManager(@NotNull String dbPrefix, @NotNull Environment environment) {
		this(dbPrefix, environment.name().toLowerCase(Locale.ROOT));
	}

	public DatabaseManager(@NotNull String dbPrefix, @NotNull String dbSuffix) {
		this.database = mongoClient.getDatabase(dbPrefix + "-" + dbSuffix);
	}

	// collectionNameOf

	private static String collectionNameOf(Object object) {
		return collectionNameOf(object.getClass());
	}

	private static String collectionNameOf(Class<?> clazz) {
		String collectionName;
		if (clazz.isAnnotationPresent(CollectionName.class))
			collectionName = clazz.getAnnotation(CollectionName.class).name();
		else
			collectionName = clazz.getSimpleName();
		return collectionName.substring(0, 127);
	}

	// collection

	public <T> @NotNull MongoCollection<T> collection(@NotNull Class<T> tClass) {
		return collection(collectionNameOf(tClass), tClass);
	}

	public <T> @NotNull MongoCollection<T> collection(@NotNull String collectionName, @NotNull Class<T> tClass) {
		return database.getCollection(collectionName, tClass);
	}

	// getAll

	public <T> @NotNull Flux<T> getAll(@NotNull Class<T> tClass) {
		return Flux.from(collection(tClass).find());
	}

	// getAllBy

	public <T> @NotNull Flux<T> getAllBy(@NotNull Bson filters, @NotNull Class<T> tClass) {
		return Flux.from(collection(tClass).find(filters));
	}

	public <T> @NotNull Flux<T> getAllByEquals(@NotNull Map<String, ?> keyValueMap, @NotNull Class<T> tClass) {
		Bson filter = Filters.empty();
		for (Map.Entry<String, ?> entry : keyValueMap.entrySet())
			filter = Filters.and(filter, Filters.eq(entry.getKey(), entry.getValue()));
		return getAllBy(filter, tClass);
	}

	// TODO: getByUUID

	// misc

	public void close() {
		mongoClient.close();
	}
}

File: db/src/main/kotlin/dev/qixils/aviary/db/DatabaseManager.kt
package dev.qixils.aviary.db

import com.mongodb.ConnectionString
import com.mongodb.MongoClientSettings
import com.mongodb.client.model.Filters
import com.mongodb.reactivestreams.client.MongoClients
import com.mongodb.reactivestreams.client.MongoCollection
import com.mongodb.reactivestreams.client.MongoDatabase
import org.bson.codecs.configuration.CodecRegistries
import org.bson.codecs.pojo.PojoCodecProvider
import org.bson.conversions.Bson
import reactor.core.publisher.Flux
import java.io.Closeable

// TODO: allow servers to add custom codecs (serializers) (i.e. for Minestom objects)
// TODO: idk if i love that this is using mongo

/**
 * Manages the connection to a MongoDB database.
 */
class DatabaseManager(dbPrefix: String, dbSuffix: String) : Closeable {
	private val pojoRegistry = CodecRegistries.fromRegistries(
		MongoClientSettings.getDefaultCodecRegistry(),
		CodecRegistries.fromProviders(PojoCodecProvider.builder().automatic(true).build())
	)
	private val mongoClientSettings = MongoClientSettings.builder()
		.applyConnectionString(ConnectionString("mongodb://localhost:27017"))
		.codecRegistry(pojoRegistry)
		.build()
	private val mongoClient = MongoClients.create(mongoClientSettings)
	private val database: MongoDatabase

	constructor(dbPrefix: String, environment: Environment) : this(
		dbPrefix,
		environment.name.lowercase()
	)

	init {
		database = mongoClient.getDatabase("$dbPrefix-$dbSuffix")
	}

	// collection
	fun <T> collection(tClass: Class<T>): MongoCollection<T> {
		return collection(collectionNameOf(tClass), tClass)
	}

	fun <T> collection(collectionName: String, tClass: Class<T>): MongoCollection<T> {
		return database.getCollection(collectionName, tClass)
	}

	// getAll
	fun <T> getAll(tClass: Class<T>): Flux<T> {
		return Flux.from(collection(tClass).find())
	}

	// getAllBy
	fun <T> getAllBy(filters: Bson, tClass: Class<T>): Flux<T> {
		return Flux.from(collection(tClass).find(filters))
	}

	fun <T> getAllByEquals(keyValueMap: Map<String?, *>, tClass: Class<T>): Flux<T> {
		var filter = Filters.empty()
		for ((key, value) in keyValueMap) filter = Filters.and(
			filter, Filters.eq(
				key, value
			)
		)
		return getAllBy(filter, tClass)
	}

	// TODO: getByUUID
	// TODO: updateBy
	// TODO: updateByUUID
	// TODO: editBy (getBy + updateBy)
	// TODO: editByUUID (getByUUID + updateByUUID)

	// misc
	override fun close() {
		mongoClient.close()
	}

	companion object {
		// collectionNameOf
		private fun collectionNameOf(obj: Any): String {
			return collectionNameOf(obj.javaClass)
		}

		private fun collectionNameOf(jClass: Class<*>): String {
			val collectionName = if (jClass.isAnnotationPresent(CollectionName::class.java))
				jClass.getAnnotation(CollectionName::class.java).name
			else
				jClass.simpleName
			return collectionName.substring(0, 127)
		}
	}
}
